enhancements/single-node/single-node-openshift-with-workers.md

```diff
@@ -0,0 +1,445 @@
+of the `IngressController`'s `.spec.replicas` and `.spec.nodePlacement`
+parameters.  In the absence of an `IngressController` resource created by the
+user/installer, or when the user/installer creates an `IngressController` with
+these two parameters omitted, the Cluster Ingress Operator will choose the
+default values for those parameters based on the value of `IngressPlacement`.
+
+If the value of `IngressPlacement` itself is omitted by the installer, it
+is defaulted to `Workers`. This would be useful to maintain the current behavior
+if the API change is merged before the installer change to set this field, or if
+we want this newer OCP version to be backwards-compatible with older installers.
+
+When the value of `IngressPlacement` is `Workers`, the defaulting behavior of
+`.spec.replicas` and `.spec.nodePlacement` will be the same as it is today:
+`.spec.replicas` will be chosen according to the value of
+`InfrastructureTopology`, namely `1` when `SingleReplica` or `2` when
+`HighlyAvailable`. `.spec.nodePlacement` will always just be:
+
+```yaml
+nodePlacement:
+  nodeSelector:
+    matchLabels:
+      kubernetes.io/os: linux
+      node-role.kubernetes.io/worker: ''
+```
+
+However, if the value of `IngressPlacement` is `ControlPlane`, the defaulting
+behavior will be different: `.spec.replicas` will be chosen instead according
+to the value of `ControlPlaneTopology`; again, `1` when `SingleReplica` or `2`
+when `HighlyAvailable`. `.spec.nodePlacement` will be always just be:
+
+```yaml
+nodePlacement:
+  nodeSelector:
+    matchLabels:
+      kubernetes.io/os: linux
+      node-role.kubernetes.io/master: ''
+```
+
+(Note that the `kubernetes.io/os: linux` label is mentioned just because it's
+the current behavior, it has no importance in this enhancement)
+
+The installer will detect situations in which it's unlikely the user will want
+to set up a load-balancer. For now, those situations only include installation
+of single control-plane node cluster deployments on the "none" platform. In
+those situations, the installer will set `IngressPlacement` to be
+`ControlPlane`. Since there's just a single control-plane node, `ControlPlane`
+topology would be `SingleReplica` and the combined effect would be that the
+`IngressController` will have just a single replica and be pinned to the single
+control-plane node. This will then ensure that the `router-default` deployment
+created by the Cluster Ingress Operator will always run on the single
+control-plane node, and as a result any `*.apps.<cluster>.<base>` DNS entries
+which originally pointed at the single control-plane node will remain correct
+even in the face of newly added worker nodes.
+
+In any other situations, the installer will set `IngressPlacement` to
+`Workers`, resulting in the same default behavior as before this enhancement,
+namely that `IngressController` pod replicas are scheduled on worker nodes and
+determined according to the `InfrastructureTopology`.
+
```

Comment:
Can the user influence the setting through the installer by asking for it to be set explicitly?

Can the user change the setting in a running cluster?

Reply:
>Can the user influence the setting through the installer by asking for it to be set explicitly?

No plans for that at the moment, but there's nothing stopping us from adding such configuration to the installer in the future, if we want users to be exposed to this

>Can the user change the setting in a running cluster?

I'm not sure about the mutability of the Ingress config CR in general. Can our API declare whether specific CR parameters are configurable, or does mutability apply to the entire CR as a whole?

I don't see a reason why users may even want to change this - users can just modify the IngressController directly to match the configuration they desire. There's no reason for them to edit a configuration whose sole purpose is to affect the defaulting behavior of new (or the first router-default) IngressController(s).

If we do make this parameter mutable, changing it should have no effect on existing IngressController CRs, since the effect of this parameter is just how IngressControllers get their parameters set by default, and since they already have been created, they already have those parameters set, so there's nothing to default.

Given this kind of unintuitive "this parameter only controls the defaulting behavior of another CR" I lean towards leaving it immutable to avoid the confusion and not give users the false impression that changing it after the cluster is up and running would have any effect on the existing IngressControllers.

Let me know what you think, and I'll write that down more concretely in the enhancement.

Reply:
> Can our API declare whether specific CR parameters are configurable

In general we should assume that any spec field is mutable. If you need immutability, this can only be achieved via a validating webhook at the moment

> I don't see a reason why users may even want to change this

It's important to note in the enhancement the impact of what would happen if they did, because the likelihood is that some user will try

Reading through this comment you've assumed fields are immutable, which has influenced your reply, is there anything you would change about your comment given all fields are mutable by default?

Reply:
>In general we should assume that any spec field is mutable. If you need immutability, this can only be achieved via a validating webhook at the moment

Good to know. I was told that we don't currently have a validating webhook for this resource, and I'm not sure this would be a good enough reason to add one.

>Reading through this comment you've assumed fields are immutable, which has influenced your reply, is there anything you would change about your comment given all fields are mutable by default?

I have not made that assumption, it doesn't have to be immutable, it's just my preference that it will be, because of its confusing natuve - I wouldn't want users to touch it.

But if it can't be trivially made immutable, then so be it - It would have to stay mutable and confusing.

Or alternatively, if we can come up with a way to solve our problems without this new API, then we wouldn't need it at all.

Reply:
> I was told that we don't currently have a validating webhook for this resource, and I'm not sure this would be a good enough reason to add one.

I suspect something like this has been said many times before, if every enhancement wants one small piece of validation, but none warrants it alone, it will never be done, but if you look wider, it would probably be worth it to start it, and then other validations may follow

> I have not made that assumption, it doesn't have to be immutable, it's just my preference that it will be, because of its confusing nature - I wouldn't want users to touch it.

Sorry, I was posing my point off of "If we do make this parameter mutable", I think if we believe the parameter is confusing, or should be immutable, there are other ways that we can do this. For example, if you were to add the parameter to the status of some object (I guess the IngressController is created by the installer and is a singleton right?), then it is _harder_ for a user to modify it. Generally spec fields are considered to be owned by the user, where status is owned by the cluster. We have used status fields on objects (eg Infrastructure) before for information that we want to have set from install time that we then prefer to be immutable later

Reply:
In general, we can't just tell users not to touch fields. If they can modify it, we need to anticipate that they will. If we want the fields to be immutable, we need to do the work to make them immutable.

The existing topology fields are good examples of fields that we want to consider immutable right now, but that might be mutable in the future. Those are in the status field of the infrastructure CR.

Reply:
Just a note on this, I'm ok with us putting this in the status to make it "immutable", but we've done this in the past for other things that we have then moved into the spec, now we have a deprecated field in the status which must be observed forever. My concern is we are creating future toil here by trying to make this immutable

I would suggest that you consider whether we can just allow the user to change the `DefaultPlacement` field value, I think given the description on the field (which is now really thorough, thank you!), we can probably say that we have documented the behaviour of the field sufficiently such that a user should understand that if it's changed, it won't affect existing `IngressController` instances. Were there any other concerns about this field being mutable?